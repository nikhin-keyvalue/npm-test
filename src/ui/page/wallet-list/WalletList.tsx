import Preact, { h } from 'preact';
import { useEffect } from 'preact/hooks';

import { getChainParams, getFallbackWalletIcon, getLocalStorageWalletConnect, parseChain } from 'ui/utils/common';
import { addChain, addMetaMaskWalletListener, connectMetaMask, switchChain } from 'ui/wallets/Metamask';
import { connectProvider, connectWalletConnect } from 'ui/wallets/WalletConnect';
import { connectPhantom } from 'ui/wallets/Phantom';
import { EthChains, MetamaskErrorCode } from 'constants/wallet';
import { ModalTypes, Wallets } from 'constants/common';
import { ConnectionInformation, WalletListProps } from 'types';
import { ASSETS_PATH } from 'config';
import { getDeviceOS } from 'zamp/payment/utils';
import { isEthChain } from 'ui/utils/wallet';
import { ConnectionResponse } from 'types/wallet';

import './style.scss';

const WalletList: Preact.FunctionComponent<WalletListProps> = ({
  setSelectedWallet,
  setAddress,
  setCurrentChain,
  selectedChainObject,
  walletList,
  setRemoveWalletListener,
  onWalletConnected,
  setModalOpen,
  rpcUrls
}) => {
  const disconnectHandler = () => {
    setAddress(null);
  };

  const accountChangeHandler = (accounts: string | any[]) => {
    setAddress(accounts.length > 0 ? (accounts[0] as string) : null);
  };

  const chainChangeHandler = (newChain: EthChains | number) => {
    setCurrentChain(parseChain(newChain));
  };

  const handleMetamaskConnect = async () => {
    const resp = await connectMetaMask();
    let isOnValidChain = true;

    if (resp)
      if (resp.address) {
        const selectedChain = selectedChainObject.network_chain_id;

        if (resp.chain !== selectedChain) {
          isOnValidChain = false;
          let addChainResp = true;
          let switchChainResp = true;

          if (selectedChain === EthChains.MUMBAI || selectedChain === EthChains.POLYGON)
            addChainResp = await addChain(getChainParams(selectedChain));
          switchChainResp = await switchChain(selectedChain);
          if (switchChainResp) {
            setCurrentChain(selectedChain);
            setAddress(resp.address);
          }
          isOnValidChain = addChainResp && switchChainResp;
        } else {
          setCurrentChain(resp.chain);
          setAddress(resp.address);
        }
        const removeWalletListener = addMetaMaskWalletListener(accountChangeHandler, chainChangeHandler);

        setRemoveWalletListener(removeWalletListener as () => void);
        setModalOpen(null);
      } else if (!resp.success) {
        if (resp.error === MetamaskErrorCode.EXISTING_REQUEST) setModalOpen(ModalTypes.METAMASK_PENDING_REQUEST);
        else setModalOpen(ModalTypes.GENERIC_ERROR);
      } else {
        setModalOpen(null);
      }
    else setModalOpen(ModalTypes.WALLET_INSTALL);

    return isOnValidChain ? resp : null;
  };

  const handlePhantomClick = async () => {
    const resp = await connectPhantom();

    if (resp)
      if (resp.address) {
        setAddress(resp.address);
        setModalOpen(null);

        return resp;
      } else if (!resp.success) {
        setModalOpen(ModalTypes.GENERIC_ERROR);
      } else {
        setModalOpen(null);
      }
    else setModalOpen(ModalTypes.WALLET_INSTALL);
  };

  const handleWalletConnectClick = async () => {
    // TODO: parse to EthChains
    const resp = await connectWalletConnect(
      selectedChainObject.network_chain_id as EthChains,
      getDeviceOS(),
      accountChangeHandler,
      chainChangeHandler,
      disconnectHandler,
      rpcUrls
    );

    if (resp.address)
      if (resp?.chain === selectedChainObject.network_chain_id) {
        if (resp.chain === selectedChainObject.network_chain_id) {
          setCurrentChain(resp.chain);
          setAddress(resp.address);
        }

        return resp;
      } else if (!resp.success) {
        // TODO: ignore or add error notifier
      } else {
        setModalOpen(ModalTypes.INVALID_CHAIN);
      }
  };

  const onWalletSelect = async (item: Wallets) => {
    setSelectedWallet(walletList[item]);

    let connectionResponse: ConnectionResponse;
    let data: ConnectionInformation;

    if (item !== Wallets.WALLETCONNECT && item !== Wallets.BINANCE) setModalOpen(ModalTypes.WALLET_CONNECT);

    switch (walletList[item].name) {
      case Wallets.METAMASK:
        connectionResponse = await handleMetamaskConnect();
        break;
      case Wallets.PHANTOM:
        connectionResponse = await handlePhantomClick();
        break;
      case Wallets.WALLETCONNECT:
        connectionResponse = await handleWalletConnectClick();
        break;
      case Wallets.BINANCE:
        data = {
          payment_option_id: walletList[item].id,
          network_code: selectedChainObject.network_code
        };
        onWalletConnected(data, walletList[item].name);
        break;
      default:
        return;
    }

    if (connectionResponse?.address) {
      data = {
        payment_option_id: walletList[item].id,
        source_account: connectionResponse.address,
        network_code: selectedChainObject.network_code
      };

      onWalletConnected(data);
    }
  };

  const handleWalletClick = async (item: Wallets) => {
    await onWalletSelect(item);
  };

  // To handle wallet side WC Disconnects, must initialize this method to clear local storage
  useEffect(() => {
    if (isEthChain(selectedChainObject.network_chain_id) && getLocalStorageWalletConnect() && rpcUrls)
      connectProvider(selectedChainObject.network_chain_id, rpcUrls);
  }, [rpcUrls]);

  return (
    <div>
      {/* TODO: tobe removed */}
      <div className='zamp-wallet-container' role='presentation' key='nikhin' onClick={() => null}>
        <div className='zamp-wallet-label-container'>
          <img src={`${ASSETS_PATH}/public/icons/generic-crypto.svg`} alt='' className='wallet-icons' />
          <div className='wallet-name'>Pay with Exchange</div>
        </div>
        <div className='arrow-container'>
          <img src={`${ASSETS_PATH}/public/icons/exchange-icons.svg`} alt='' height='15' style='margin-right: 10px;' />
          <img src={`${ASSETS_PATH}/public/images/right-arrow.svg`} alt='' height='11' />
        </div>
      </div>
      <div className='zamp-wallet-container' role='presentation' key='nik' onClick={() => null}>
        <div className='zamp-wallet-label-container'>
          <img src={`${ASSETS_PATH}/public/icons/pay-with-qr.svg`} alt='' className='wallet-icons' />
          <div className='wallet-name'>Pay with QR</div>
        </div>
        <div className='arrow-container'>
          <img src={`${ASSETS_PATH}/public/images/right-arrow.svg`} alt='' height='11' />
        </div>
      </div>
      {/* Till here */}

      {selectedChainObject?.eligible_payment_options?.map((item) => {
        return (
          <div className='zamp-wallet-container' role='presentation' key={item} onClick={() => handleWalletClick(item)}>
            <div className='zamp-wallet-label-container'>
              <img
                src={walletList[item]?.logo_url}
                alt='wallet-icon'
                className='wallet-icons'
                onError={(e) => {
                  e.currentTarget.onerror = null;
                  e.currentTarget.src = getFallbackWalletIcon(item);
                }}
              />
              <div className='wallet-name'>{walletList[item]?.name}</div>
            </div>
            <div className='arrow-container'>
              <img src={`${ASSETS_PATH}/public/images/right-arrow.svg`} alt='' height='11' />
            </div>
          </div>
        );
      })}
    </div>
  );
};

export default WalletList;
