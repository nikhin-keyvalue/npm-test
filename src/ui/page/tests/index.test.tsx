// TODO: remove
/* eslint-disable @typescript-eslint/no-unused-vars */
import { h } from 'preact';
import { configure, mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Adapter } from 'enzyme-adapter-preact-pure';

import WalletList from 'ui/page/wallet-list/WalletList';

import Content from '../index';

configure({ adapter: new Adapter() });

const zampSession = {
  paymentInfo: {
    id: 1,
    source_amount: 0
  },
  merchantConfig: {
    merchant: {},
    payment_options: [],
    available_networks: [{ network_code: 'ETH', eligible_payment_options: [] }],
    default_network: 'ETH'
  }
} as any;
// TODO: to be typed

const paymentSession = {} as any;

describe('Content', () => {
  const onPaymentCompletedMock = jest.fn();
  const onPaymentCancelMock = jest.fn();
  const mockFn = jest.fn();

  it('render / unmount page', () => {
    const wrapper = mount(
      <Content
        onPaymentCompleted={onPaymentCompletedMock}
        onPaymentCancel={onPaymentCancelMock}
        zampSession={zampSession}
        paymentSession={paymentSession}
        isError={false}
      />
    );
    const tree = shallowToJson(wrapper);

    expect(tree).toMatchSnapshot();
    wrapper.unmount();
  });

  it('Mount WalletList and handle setRemoveWalletListener', () => {
    const wrapper = shallow(
      <Content
        onPaymentCompleted={onPaymentCompletedMock}
        onPaymentCancel={onPaymentCancelMock}
        zampSession={zampSession}
        paymentSession={paymentSession}
        isError={false}
      />
    );
    const removeListenerMock = () => {};

    // wrapper.find(WalletList).props().setRemoveWalletListener(removeListenerMock);
  });

  it('render Coin List through WalletList', () => {
    const wrapper = shallow(
      <Content
        onPaymentCompleted={mockFn}
        onPaymentCancel={mockFn}
        zampSession={zampSession}
        paymentSession={paymentSession}
        isError={false}
      />
    );

    // wrapper.find(WalletList).props().handlePageChange(ZampfiPages.COIN_LIST);
  });

  it('render Success Screen through CoinList', () => {
    const wrapper = shallow(
      <Content
        onPaymentCompleted={mockFn}
        onPaymentCancel={mockFn}
        zampSession={zampSession}
        paymentSession={paymentSession}
        isError={false}
      />
    );

    // wrapper.find(WalletList).props().handlePageChange(ZampfiPages.COIN_LIST);
    // wrapper.find(CoinList).props().handlePageChange(ZampfiPages.SUCCESS_SCREEN);
  });

  it('render Error Screen through CoinList', () => {
    const wrapper = shallow(
      <Content
        onPaymentCompleted={mockFn}
        onPaymentCancel={mockFn}
        zampSession={zampSession}
        paymentSession={paymentSession}
        isError={false}
      />
    );

    // wrapper.find(WalletList).props().handlePageChange(ZampfiPages.COIN_LIST);
    // wrapper.find(CoinList).props().handlePageChange(ZampfiPages.ERROR_SCREEN);
  });

  it('render default invalid case', () => {
    const wrapper = shallow(
      <Content
        onPaymentCompleted={mockFn}
        onPaymentCancel={mockFn}
        zampSession={zampSession}
        paymentSession={paymentSession}
        isError={false}
      />
    );

    // wrapper.find(WalletList).props().handlePageChange('');
  });

  it('handle onPaymentComplete success, error, and invalid cases', () => {
    const wrapper = shallow(
      <Content
        onPaymentCompleted={onPaymentCompletedMock}
        onPaymentCancel={onPaymentCancelMock}
        zampSession={zampSession}
        paymentSession={paymentSession}
        isError={false}
      />
    );

    // wrapper.find(WalletList).props().handlePageChange(ZampfiPages.COIN_LIST);
    // wrapper.find(CoinList).props().handlePaymentCompleted('success');
    // wrapper.find(CoinList).props().handlePaymentCompleted(); // error case
    // wrapper.find(CoinList).props().handlePaymentCompleted('invalid');
  });
});
