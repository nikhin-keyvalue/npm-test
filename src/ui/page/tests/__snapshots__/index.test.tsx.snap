// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content render / unmount page 1`] = `
<Content
  isError={false}
  onPaymentCancel={[MockFunction]}
  onPaymentCompleted={[MockFunction]}
  paymentSession={Object {}}
  zampSession={
    Object {
      "merchantConfig": Object {
        "available_networks": Array [
          Object {
            "eligible_payment_options": Array [],
            "network_code": "ETH",
          },
        ],
        "default_network": "ETH",
        "merchant": Object {},
        "payment_options": Array [],
      },
      "paymentInfo": Object {
        "id": 1,
        "source_amount": 0,
      },
    }
  }
>
  <div
    className="zamp-checkout-page"
    onClick={[Function]}
    role="presentation"
  >
    <Topbar
      isDisabled={false}
      onBackButtonClick={[Function]}
    >
      <div
        className="zamp-topbar-wrapper"
      >
        <div
          className="zamp-topbar-go-back-wrapper"
        >
          <img
            alt="Back button"
            className="zamp-topbar-back-icon false"
            height="12"
            onClick={[Function]}
            src="undefined/public/icons/back.svg"
          />
        </div>
        <div
          className="zamp-topbar-title"
        >
          Recharge by crypto
        </div>
      </div>
    </Topbar>
    <div
      className="zamp-checkout-page-contents"
    >
      <div
        className="zamp-checkout-page-content-body"
      >
        <Header
          orderId={1}
          sourceAmount={0}
          userDetails={Object {}}
        >
          <div
            className="zamp-header-wrapper"
          >
            <div
              className="zamp-header-title"
            >
              <img
                alt="Merchant logo"
                height="30"
              />
            </div>
            <div
              className="zamp-amount-container"
            >
              <div
                className="zamp-amount"
              >
                 
                0
                <span
                  className="zamp-amount-decimal-value"
                >
                  .00
                </span>
              </div>
            </div>
          </div>
        </Header>
        <div
          className="zamp-content-wrapper false"
        >
          <div
            className="zamp-title"
          >
            Network
          </div>
          <NetworkSelect
            isBottomSheetOpen={null}
            isDropDownActive={false}
            itemInitialValue={
              Object {
                "eligible_payment_options": Array [],
                "network_code": "ETH",
              }
            }
            itemsList={
              Array [
                Object {
                  "eligible_payment_options": Array [],
                  "network_code": "ETH",
                },
              ]
            }
            onSelectItem={[Function]}
            setBottomSheetOpen={[Function]}
            setDropDownActive={[Function]}
          >
            <div
              className="zamp-dropdown-wrapper"
            >
              <div
                className="zamp-dropdown-list "
              />
            </div>
          </NetworkSelect>
          <div
            className="zamp-title"
          >
            Wallet/Exchange
          </div>
          <WalletList
            onWalletConnected={[Function]}
            rpcUrls={Map {}}
            selectedChainObject={
              Object {
                "eligible_payment_options": Array [],
                "network_code": "ETH",
              }
            }
            setAddress={[Function]}
            setCurrentChain={[Function]}
            setModalOpen={[Function]}
            setRemoveWalletListener={[Function]}
            setSelectedWallet={[Function]}
            walletList={Array []}
          >
            <div>
              <div
                className="zamp-wallet-container"
                key="nikhin"
                onClick={[Function]}
                role="presentation"
              >
                <div
                  className="zamp-wallet-label-container"
                >
                  <img
                    alt=""
                    className="wallet-icons"
                    src="undefined/public/icons/generic-crypto.svg"
                  />
                  <div
                    className="wallet-name"
                  >
                    Pay with Exchange
                  </div>
                </div>
                <div
                  className="arrow-container"
                >
                  <img
                    alt=""
                    height="15"
                    src="undefined/public/icons/exchange-icons.svg"
                    style="margin-right: 10px;"
                  />
                  <img
                    alt=""
                    height="11"
                    src="undefined/public/images/right-arrow.svg"
                  />
                </div>
              </div>
              <div
                className="zamp-wallet-container"
                key="nik"
                onClick={[Function]}
                role="presentation"
              >
                <div
                  className="zamp-wallet-label-container"
                >
                  <img
                    alt=""
                    className="wallet-icons"
                    src="undefined/public/icons/pay-with-qr.svg"
                  />
                  <div
                    className="wallet-name"
                  >
                    Pay with QR
                  </div>
                </div>
                <div
                  className="arrow-container"
                >
                  <img
                    alt=""
                    height="11"
                    src="undefined/public/images/right-arrow.svg"
                  />
                </div>
              </div>
            </div>
          </WalletList>
        </div>
      </div>
      <Footer>
        <div
          className="zamp-footer-wrapper"
        >
          powered by 
          <span
            className="zamp-footer-name"
          >
            ZAMP
          </span>
        </div>
      </Footer>
    </div>
    <BottomSheet
      isOpen={false}
      setBottomSheetOpen={[Function]}
    >
      <div
        className="zamp-bottom-sheet-wrapper "
        onClick={[Function]}
        role="presentation"
      >
        <div
          className="zamp-bottom-sheet-content zamp-sheet-content-close "
          onClick={[Function]}
          role="presentation"
        >
          <GoBack
            description=" upon confirmation, this transaction will be cancelled, and you’ll have to start a new payment"
            onPrimaryButtonClick={[Function]}
            primaryButtonLabel="Yes, leave"
            secondaryButtonLabel="No"
            title="Leaving? you won’t be able to continue this transaction"
            toggleSheetOpen={[Function]}
          >
            <div>
              <div
                className="zamp-bottom-sheet-title"
              >
                Leaving? you won’t be able to continue this transaction
              </div>
              <div
                className="zamp-bottom-sheet-desc"
              >
                 upon confirmation, this transaction will be cancelled, and you’ll have to start a new payment
              </div>
              <div
                className="zamp-bottom-sheet-button-container"
              >
                <Button
                  buttonLabel="Yes, leave"
                  handleButtonClick={[Function]}
                  isDisabled={false}
                  type="primary"
                >
                  <button
                    className="zamp-primary-button"
                    disabled={false}
                    onClick={[Function]}
                    style={
                      Object {
                        "height": undefined,
                        "width": undefined,
                      }
                    }
                  >
                    Yes, leave
                  </button>
                </Button>
                <Button
                  buttonLabel="No"
                  handleButtonClick={[Function]}
                  isDisabled={false}
                  type="secondary"
                >
                  <button
                    className="zamp-secondary-button"
                    disabled={false}
                    onClick={[Function]}
                    style={
                      Object {
                        "height": undefined,
                        "width": undefined,
                      }
                    }
                  >
                    No
                  </button>
                </Button>
              </div>
            </div>
          </GoBack>
        </div>
      </div>
    </BottomSheet>
  </div>
</Content>
`;
